### PR TITLE
Store: Split dashboard rendering into setup or normal mode

### LINKED ISSUE #19481: Store: Re-Work Dashboard Logic for NUX
Today while testing out the new store nux flow, @lamosty discovered some issues with the Dashboard page during the sign-up process that prevented the loading/status indicator screens from working.

@allendav and I ran some further tests and initially thought the problem may have been a result of the new Dashboard placeholder ( introduced in #19203 ) - so we removed that logic in #19475. Subsequent tests still resulted in various issues, so the problem appears to be a bit deeper.

Some notes form slack discussion:

> at the tail end of setting up a new site using the new /start flow, we just to a point where setup stops progressing in the browse

> the issue is (afaik) that we are using the Woo setup page to track the progress of background AT flow

Essentially the Dashboard render logic needs to better handle this situation - perhaps with a different render method for transferring sites vs established sites.

More discussion and thoughts at p1509729580000298-slack-store-on-wpcom

### PR DESCRIPTION
**Pull request by timmyc**
Fixes #19481

This branch seeks to clear up some of the issues encountered during the testing of the new store signup flow. The first step I took was to create two different sub-render methods... one for a store that is still going through setup, and one for the normal dashboard.

Setup is determined by the following selectors: `getFinishedInitialSetup`, `getFinishedInstallOfRequiredPlugins`, `getSetStoreAddressDuringInitialSetup`, and `getFinishedPageSetup`. If any of these are falsey, the setup dashboard is rendered.

Additionally during testing I discovered that `areProductsLoaded` were stuck in a perpetually loading state. This seemed to be another remnant of the switch of products to using the data layer in #19410. The fix for this here was to properly pass in a `failureAction` to the `fetchProducts` call... since I'm still green when it comes to the data layer I'm going to /cc @ryelle  on this part of the PR specifically.

__To Test__
Since this PR touches the dashboard, really all of the random states the dashboard handles should ideally be tested. I have tested close to 20 new sites today and am feeling pretty confident in the PR... and I have also tested many of these sites in various stages of the setup wizard, and sites that are already live/configured.

Specifically though, we could really use more 👀  on the entire NUX flow:

**Before you begin testing, ensure you have the following A/B tests configured properly:**
store_ _site_title_ _wordpress_com
_signupAtomicStoreVsPressable_ set to `atomic`

store_ _site_title_ _wordpress_com
_gsuiteUpsell_ to `hide` **This is super important** in my testing the flow completely fails when the gsuite upsell is shown ( over 50% failure rate ) - I will create a PR to disable the gsuiteUpsell if the Atomic test is true.

Also give yourself some credits... this will get expensive.

__To Actually Test__
Now that you have done the A/B test configurations above you can now:

- Visit http://calypso.localhost:3000/start/design-type-with-store-nux
- Use a low-value .blog domain during the process
- Ultimately verify that you end up on the Store Dashboard and see a loading indicator which displays the progress of the site/store being setup
- When the Store Address field is shown, that is a _great success_
- From there play with the various states of the dashboard setup wizard and make sure nothing has regressed there.

### CODE REVIEW

**Comment by alisterscott** [comment]
I completed a new sign up alisterscottteststoretesting.blog with great success, the first sign up atomic store I was able to complete so well done on this. I didn't come across any issues.

File: client/extensions/woocommerce/app/dashboard/index.js [hunk 1]
```diff
@@ -50,49 +50,61 @@ class Dashboard extends Component {
...
 			this.props.requestSettings( newSiteId );
 
 			if ( ! productsLoaded ) {
-				this.props.fetchProducts( newSiteId, { page: 1 } );
+				this.fetchSiteProducts( newSiteId );
 			}
 		}
 	};
 
+	fetchSiteProducts = siteId => {
```
Comment (budzanowski) [inline client/extensions/woocommerce/app/dashboard/index.js]: Nicely extracted.

File: client/extensions/woocommerce/app/dashboard/index.js [hunk 2]
```diff
@@ -50,49 +50,61 @@ class Dashboard extends Component {
...
 
 	componentDidMount = () => {
-		const { selectedSite, productsLoaded } = this.props;
+		const { siteId, productsLoaded } = this.props;
 
-		if ( selectedSite && selectedSite.ID ) {
-			this.props.fetchSetupChoices( selectedSite.ID );
-			this.props.fetchOrders( selectedSite.ID );
-			this.props.requestSettings( selectedSite.ID );
+		if ( siteId ) {
```
Comment (budzanowski) [inline client/extensions/woocommerce/app/dashboard/index.js]: I think that this could be extracted. Fetching here a and in componentWillReceiveProps is the same.

File: client/extensions/woocommerce/app/dashboard/index.js [hunk 3]
```diff
@@ -163,53 +183,62 @@ class Dashboard extends Component {
...
 	const hasOrders = getNewOrdersWithoutPayPalPending( state ).length > 0;
 	const hasProducts = getTotalProducts( state ) > 0;
 	const productsLoaded = areProductsLoaded( state );
 	const finishedInitialSetup = getFinishedInitialSetup( state );
+	const finishedInstallOfRequiredPlugins = getFinishedInstallOfRequiredPlugins( state );
+	const finishedPageSetup = getFinishedPageSetup( state );
+	const setStoreAddressDuringInitialSetup = getSetStoreAddressDuringInitialSetup( state );
+
+	// This prop toggles the display of the setup steps vs the normal dashboard.
+	const isSetupComplete =
```
Comment (budzanowski) [inline client/extensions/woocommerce/app/dashboard/index.js]: This could be a selector. It would nicely encapsulate logic.

File: client/extensions/woocommerce/app/dashboard/index.js [hunk 4]
```diff
@@ -50,49 +50,61 @@ class Dashboard extends Component {
...
+				this.fetchSiteProducts( newSiteId );
 			}
 		}
 	};
 
+	fetchSiteProducts = siteId => {
+		const params = { page: 1 };
+		const failureAction = dispatch => {
+			dispatch( this.props.fetchProductsFailure( siteId, params ) );
+		};
```
Comment (jsnajdr) [inline client/extensions/woocommerce/app/dashboard/index.js]: This is very strange. `this.props.fetchProductsFailure` is a bound action creator. That means calling it creates the action and immediately dispatches it.

Calling `dispatch` generally returns the just-dispatched action object (if not intercepted by some middleware). Your explicit `dispatch` call therefore dispatches the failure action for a second time.

I think what you really want here is just create the failure action and pass it as parameter to `fetchProducts`:
```js
// call the imported action creator. It creates an action object.
// It's all just data, nothing gets executed.
const failureAction = fetchProductsFailure( siteId, params );
// call the bound action creator. This calls `dispatch` internally and executes the request.
this.props.fetchProducts( siteId, params, null, failureAction );
```

File: client/extensions/woocommerce/app/dashboard/index.js [hunk 5]
```diff
@@ -219,6 +248,7 @@ function mapDispatchToProps( dispatch ) {
 			fetchOrders,
 			fetchSetupChoices,
 			fetchProducts,
+			fetchProductsFailure,
```
Comment (jsnajdr) [inline client/extensions/woocommerce/app/dashboard/index.js]: There's no need to bind the `fetchProductFailure` action creator here.

File: client/extensions/woocommerce/app/dashboard/index.js [hunk 6]
```diff
@@ -50,49 +50,61 @@ class Dashboard extends Component {
...
 		}
 	};
 
+	fetchSiteProducts = siteId => {
+		const params = { page: 1 };
+		const failureAction = dispatch => {
+			dispatch( this.props.fetchProductsFailure( siteId, params ) );
+		};
+
+		this.props.fetchProducts( siteId, params, null, failureAction );
```
Comment (jsnajdr) [inline client/extensions/woocommerce/app/dashboard/index.js]: We want to set the `isLoading` state flag to `false` every time the `fetchProducts` request fails, no matter where it's called from, don't we? In that case the `failureAction` should be dispatched inside the data layer, not in app code.

The best place to dispatch this is the `apiError` function in `state/data-layer/products/index.js`.

Setting the `isLoading` when the request is initiated and when it succeeds is also performed inside data layer. The caller of `fetchProducts` doesn't need to have the responsibility for doing that.

By the way, the `fetchProducts` action creator also has a `successAction` parameter, but it's never used, even when I pass a non-null value. That's an unrelated bug report for @coderkevin

File: client/extensions/woocommerce/app/dashboard/index.js (hunk 2 above)
Comment (jsnajdr) [reply client/extensions/woocommerce/app/dashboard/index.js]: I agree with @budzanowski -- it would be nice to extract this into a method. And instead of `componentWillReceiveProps`, we can use `componentDidUpdate`. The benefit is that we can use `this.props` in both cases with the same meaning.

File: client/extensions/woocommerce/app/dashboard/index.js (hunk 6 above)
Comment (timmyc) [reply client/extensions/woocommerce/app/dashboard/index.js]: Indeed, I agree - and had actually started putting the logic there. I will refactor this a bit and add some tests too.

**Comment by timmyc** [comment]
@jsnajdr and @budzanowski thank you much for the reviews. I have addressed the feedback and would love to have you both check this out again. In 4e0188a I created a new combined `isStoreSetupComplete` selector with some test coverage. e6aa088 adds an auto-dispatched WOOCOMMERCE_PRODUCTS_REQUEST_FAILURE to the data layer when the API returns an error - thus removing the burden from the calling component. And finally in 4beada8, I refactored the lifecycle methods a bit further to remove duplication.

**Comment by timmyc** [comment]
@jsnajdr and @budzanowski thank you much for the reviews. I have addressed the feedback and would love to have you both check this out again. In 4e0188a I created a new combined `isStoreSetupComplete` selector with some test coverage. e6aa088 adds an auto-dispatched WOOCOMMERCE_PRODUCTS_REQUEST_FAILURE to the data layer when the API returns an error - thus removing the burden from the calling component. And finally in 4beada8, I refactored the lifecycle methods a bit further to remove duplication.

**Comment by budzanowski** [comment]
Looks great after updates.

File: client/extensions/woocommerce/app/dashboard/index.js [hunk 7]
```diff
@@ -163,53 +173,55 @@ class Dashboard extends Component {
 		return (
 			<div>
 				{ manageView }
-				{ ! this.props.mailChimpConfigured &&
-					( config.isEnabled( 'woocommerce/extension-settings-email' ) && (
-						<MailChimp site={ selectedSite } redirectToSettings dashboardView />
-					) ) }
+				{ ! this.props.mailChimpConfigured && (
+					<MailChimp site={ selectedSite } redirectToSettings dashboardView />
+				) }
```
Comment (jsnajdr) [inline client/extensions/woocommerce/app/dashboard/index.js]: This is an unrelated change where a `config.isEnabled` check is being removed. If you and @budzanowski can confirm that this change is wanted, this PR is ready to merge.

Comment (timmyc) [reply client/extensions/woocommerce/app/dashboard/index.js]: I confirmed with @budzanowski via DM that he is okay with this change - thanks for asking about it though!
